# NovaPay Fraudulent Transaction Detection

## Week 1 — Day 1: Foundations and Initial Setup

**Task:** Week 1: Foundations & Initial Setup  
**Project Duration:** 10 Days  
**Project Type:** Supervised Binary Classification  
**Target Variable:** `is_fraud`

---

### Day 1 Aim

The aim of Day 1 is to understand the business problem, define the project objectives and success criteria, identify the main stakeholders, and confirm that the dataset can be loaded successfully.

## 1. What Problem Are My Solving?

NovaPay is a digital money-transfer company that allows customers to send, receive and hold money across different countries.

The company faces fraud risks such as:

- identity theft;
- account takeover;
- unauthorised payments;
- transaction laundering.

NovaPay currently uses fixed fraud rules and manual reviews. These methods may miss new fraud patterns and may also flag genuine customers incorrectly.

Therefore, the business needs a machine-learning system that can identify suspicious transactions more effectively.

## 2. Project Objective

The main objective is to build a machine-learning model that can classify each transaction as either legitimate or fraudulent.

The target column is `is_fraud`:

- `0` means a legitimate transaction.
- `1` means a fraudulent transaction.

The project also aims to:

- improve fraud recall by at least 15% compared with the existing rules-based baseline;
- maintain acceptable precision so that fewer genuine customers are wrongly flagged;
- explain why a transaction is predicted as fraud;
- prepare the model for real-time transaction scoring.

## 3. Machine-Learning Problem

This is a **supervised binary classification** problem.

It is **supervised** because the dataset contains known fraud labels that the model can learn from.

It is **binary classification** because the model predicts one of two classes:

- legitimate transaction;
- fraudulent transaction.

## 4. Success Criteria

The model will mainly be evaluated using:

- **Recall:** how many actual fraudulent transactions the model correctly detects.
- **Precision:** how many transactions predicted as fraud are truly fraudulent.
- **F1-score:** the balance between recall and precision.
- **ROC-AUC:** how well the model separates fraudulent and legitimate transactions.

Accuracy will also be checked, but it will not be used alone because fraud datasets are usually imbalanced.

### Business Meaning

High recall helps NovaPay reduce missed fraud and financial loss.

High precision helps NovaPay avoid delaying genuine customers and reduces unnecessary manual reviews.

## 5. Main Stakeholders

### Fraud Analysts
They need accurate fraud alerts and fewer unnecessary manual reviews.

### Customers
They expect safe and fast transfers without incorrect fraud blocks.

### Compliance and Regulatory Teams
They need fraud decisions to be clear, explainable and auditable.

### Business Management
Management wants to reduce fraud losses, chargebacks and compliance penalties.

### Data Science and Engineering Teams
They are responsible for building, testing, deploying and maintaining the system.

## 6. Project Scope

This project will include:

- data assessment;
- data cleaning and preparation;
- feature engineering and exploratory data analysis;
- baseline and advanced machine-learning models;
- class-imbalance handling;
- model evaluation and explainability;
- API packaging;
- monitoring, drift detection and a retraining plan;
- final report, slides and handover.

## 7. Assumptions, Risks and Constraints

### Assumptions

- The `is_fraud` column contains reliable fraud labels.
- Historical transaction patterns can help predict future fraud.
- The available customer, transaction and device information is useful for fraud detection.

### Risks

- Missing or incorrect data.
- A large difference between the number of legitimate and fraudulent transactions.
- False positives that affect genuine customers.
- False negatives that allow fraud to go undetected.
- Fraud patterns changing over time.

### Constraints

- The project must be completed within 10 days.
- The final model should support real-time scoring.
- Fraud predictions should be explainable.
- Recall should improve by at least 15% compared with the rules-based baseline.

# Initial Project Setup

## 8. Import the Basic Libraries

### What problem are my solving?

Before working with the dataset, the Python libraries needed for data analysis must be imported.

- `pandas` is used to work with tables and datasets.
- `numpy` supports numerical operations.
- `matplotlib` is used for charts later in the project.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Libraries imported successfully.")

Libraries imported successfully.


## 9. Load the NovaPay Dataset

### What problem are my solving?

The Excel file must be loaded into Python so that it can be examined and prepared during the remaining project days.

In [2]:
file_name = "nova_pay_combined (1).xlsx"

df = pd.read_excel(file_name)

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 10. Confirm the Dataset Setup

This is only a basic setup check. The detailed data-quality assessment will be completed on Day 2.

In [3]:
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])
print("Target column available:", "is_fraud" in df.columns)

Number of rows: 11400
Number of columns: 26
Target column available: True


## 11. Preview the Dataset

Displaying the first five rows confirms that the dataset was imported correctly.

In [4]:
df.head()

,transaction_id,customer_id,timestamp,home_country,source_currency,dest_currency,channel,amount_src,amount_usd,fee,exchange_rate_src_to_dest,device_id,new_device,ip_address,ip_country,location_mismatch,ip_risk_score,kyc_tier,account_age_days,device_trust_score,chargeback_history_count,risk_score_internal,txn_velocity_1h,txn_velocity_24h,corridor_risk,is_fraud
0,fee8542d-8ee6-4b0d-9671-c294dd08ed26,402cccc9-28de-45b3-9af7-cc5302aa1f93,2022-10-03 18:40:59.468549+00:00,US,USD,CAD,ATM,278.19,278.19,4.25,1.351351,9f292dcc-3297-4947-a260-6a1ef69041ff,False,221.78.171.180,US,False,0.123,standard,263,0.522,0,0.223,0,0,0.0,0
1,bfdb9fc1-27fe-4a85-b043-4d813d679259,67c2c6b3-ef0a-4777-a3f1-c84a851bb6ad,2022-10-03 20:39:38.468549+00:00,CA,CAD,MXN,web,208.51,154.29,4.24,12.758621,3a95b9f5-309f-4684-a46d-e2ff2435bf78,True,120.12.20.29,CA,False,0.569,standard,947,0.475,0,0.268,0,1,0.0,0
2,fc855034-3ea5-4993-9afa-b511d93fe5e8,6d0d9b27-fa26-45f8-93b1-2df29d182d9c,2022-10-03 23:02:43.468549+00:00,US,USD,CNY,mobile,160.33,160.33,2.70,7.142857,a4737752-9aac-43ed-9d8b-2ccdffc24052,False,223.96.181.93,US,False,0.437,enhanced,367,0.939,0,0.176,0,0,0.0,0
3,2cf8c08e-42ec-444d-a755-34b9a2a0a4ca,7bd5200c-5d19-44f0-9afe-8b339a05366b,2022-10-04 01:08:53.468549+00:00,US,USD,EUR,mobile,59.41,59.41,2.22,0.925926,6aeb85a3-5603-4221-896c-9e6882764f1a,False,186.228.15.74,US,False,0.594,standard,147,0.551,0,0.391,0,0,0.0,0
4,d907a74d-b426-438d-97eb-dbe911aca91c,70a93d26-8e3a-4179-900c-a4a7a74d08e5,2022-10-04 09:35:03.468549+00:00,US,USD,INR,mobile,200.96,200.96,3.61,83.333333,a5b9250e-dbe0-4c5f-a6e7-5492b7349402,False,11.82.47.62,US,False,0.121,enhanced,257,0.894,0,0.257,0,0,0.0,0


# Day 1 Summary and Business Decision

## What Was Completed Today?

- The NovaPay business problem was defined.
- The project objective and machine-learning problem were identified.
- The success criteria and evaluation metrics were explained.
- The main stakeholders were identified.
- The project scope, assumptions, risks and constraints were recorded.
- The required Python libraries were imported.
- The NovaPay dataset was loaded successfully.
- The dataset was confirmed to contain **11,400 rows**, **26 columns** and the target variable `is_fraud`.

## Day 1 Business Decision

The dataset contains the required fraud target, so the project can proceed as a supervised binary-classification task.

The next step is to complete a detailed data assessment on Day 2 before cleaning or modelling begins.

# Day 1 Deliverable

The official Day 1 deliverable is:

**`01_kickoff_summary.md`**

It contains:

- the company background;
- the business problem;
- the project objective;
- the machine-learning problem;
- the success criteria;
- the stakeholders;
- the project scope;
- the assumptions, risks and constraints;
- the expected outcome;
- the Day 1 setup result.

## Personal Learning Reflection

Today, I learned that a machine-learning project should begin with a clear business problem before any model is built. I also learned why recall and precision are both important in fraud detection. Finally, I successfully imported the required libraries and loaded the NovaPay dataset.

# WEEK 1 – DAY 2

## Data Sources & Initial Assessment

### Objective

To understand the available NovaPay transaction data, fraud-label distribution, data-quality problems and limitations before cleaning or modelling.

### Tasks / Activities

- Explore the raw transaction dataset and its metadata fields.
- Check the confirmed fraud labels and customer chargeback history.
- Identify missing values, duplicate records and incorrect data types.
- Examine whether the fraud classes are balanced or imbalanced.
- Document the main data-quality issues and dataset limitations.

### Research

Bias and fairness in financial machine-learning systems.

### Expected Output

A data-assessment report named `02_data_assessment.md`, summarising the data scope, fraud-label distribution, data-quality problems, class imbalance and limitations.

# WEEK 1 – DAY 2

## Data Sources & Initial Assessment

### Objective

To understand the available NovaPay transaction data, fraud labels, data-quality problems and dataset limitations before cleaning or building a machine-learning model.

## Reload the Dataset After Kernel Restart

The dataset was assessed initially during Day 1. It is being reloaded because the Jupyter kernel restarted and the `df` variable was no longer available.

In [6]:
import pandas as pd

df = pd.read_excel("nova_pay_combined (1).xlsx")

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 1. Examine the Fraud-Label Distribution

### What question are my answering?

I want to find out how many transactions are legitimate and how many are fraudulent.

This is important because a large difference between the two groups may cause a machine-learning model to favour the majority class and overlook fraudulent transactions.

In [18]:
fraud_counts = df["is_fraud"].value_counts().sort_index()

fraud_percentages = (
    df["is_fraud"]
    .value_counts(normalize=True)
    .sort_index()
    .mul(100)
    .round(2)
)

print("Number of transactions:")
print(fraud_counts)

print("\nPercentage of transactions:")
print(fraud_percentages)

Number of transactions:
is_fraud
0    10403
1      997
Name: count, dtype: int64

Percentage of transactions:
is_fraud
0    91.25
1     8.75
Name: proportion, dtype: float64


### Observation

The dataset contains 10,403 legitimate transactions and 997 fraudulent transactions.

Legitimate transactions represent 91.25% of the dataset, while fraudulent transactions represent 8.75%.

This shows that the dataset is imbalanced because legitimate transactions are much more common than fraudulent transactions. This imbalance is important because a machine-learning model may focus more on the majority class and fail to detect enough fraud cases.

## 2. Examine Chargeback History and Fraud

### What question are my answering?

I want to find out whether transactions linked to customers with previous chargebacks are more likely to be fraudulent.

This is important because chargeback history may be a useful warning sign when detecting fraud.

In [8]:
chargeback_summary = pd.crosstab(
    df["chargeback_history_count"],
    df["is_fraud"]
)

chargeback_summary.columns = ["Legitimate", "Fraud"]

chargeback_summary["Total"] = (
    chargeback_summary["Legitimate"]
    + chargeback_summary["Fraud"]
)

chargeback_summary["Fraud Rate (%)"] = (
    chargeback_summary["Fraud"]
    / chargeback_summary["Total"]
    * 100
).round(2)

chargeback_summary

,Legitimate,Fraud,Total,Fraud Rate (%)
chargeback_history_count,,,,
0,10378,580,10958,5.29
1,24,307,331,92.75
2,1,110,111,99.10


### Observation

The results show a strong relationship between previous chargebacks and fraud.

For transactions with no previous chargeback, 5.29% were fraudulent.

For transactions with one previous chargeback, 92.75% were fraudulent.

For transactions with two previous chargebacks, 99.10% were fraudulent.

This suggests that `chargeback_history_count` may be an important feature for detecting fraud.

However, NovaPay must confirm that the chargeback history was already available before the transaction was classified. If the information was added after fraud was confirmed, using it in the model could cause data leakage.

## 3. Assess Missing Values

### What question are am answering?

I want to identify which columns contain missing information and measure how much data is missing.

This is important because missing values can affect analysis and machine-learning performance if they are not handled correctly.

In [9]:
missing_values = df.isnull().sum()

missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Percentage (%)": (
        missing_values / len(df) * 100
    ).round(2)
})

missing_summary = missing_summary[
    missing_summary["Missing Values"] > 0
]

missing_summary

,Missing Values,Percentage (%)
timestamp,29,0.25
amount_usd,305,2.68
fee,295,2.59
ip_address,305,2.68
ip_country,302,2.65
kyc_tier,302,2.65
device_trust_score,295,2.59


### Observation

Seven columns contain missing values.

- `timestamp` has 29 missing values, representing 0.25%.
- `amount_usd` has 305 missing values, representing 2.68%.
- `fee` has 295 missing values, representing 2.59%.
- `ip_address` has 305 missing values, representing 2.68%.
- `ip_country` has 302 missing values, representing 2.65%.
- `kyc_tier` has 302 missing values, representing 2.65%.
- `device_trust_score` has 295 missing values, representing 2.59%.

The amount of missing data is relatively small, but these values still need to be handled before modelling.

No missing values will be removed or replaced during Day 2 because data cleaning will be completed on Day 3.

## 4. Check for Duplicate Transactions

### What question are my answering?

I want to find out whether the same transaction appears more than once in the dataset.

Duplicate transactions can distort the analysis because the same transaction may be counted repeatedly. They can also affect the machine-learning model by giving repeated records too much influence.

In [10]:
duplicate_rows = df.duplicated().sum()
unique_transaction_ids = df["transaction_id"].nunique()

print("Total rows:", len(df))
print("Exact duplicate rows:", duplicate_rows)
print("Unique transaction IDs:", unique_transaction_ids)

Total rows: 11400
Exact duplicate rows: 200
Unique transaction IDs: 11200


### Observation

The dataset contains 11,400 rows but only 11,200 unique transaction IDs.

There are 200 exact duplicate rows, meaning some transactions appear more than once in the dataset.

Duplicate records can distort the analysis and may cause the machine-learning model to give repeated transactions too much importance.

The duplicate rows will be removed during Day 3 when the data-cleaning process begins.

## 5. Check Categorical Data Quality

### What question am I answering?

I want to check whether categories such as transaction channel and KYC tier are written consistently.

Spelling mistakes and differences in capital letters may cause the same category to be treated as several different categories.

In [11]:
print("Transaction channels:")
print(df["channel"].value_counts(dropna=False))

print("\nKYC tiers:")
print(df["kyc_tier"].value_counts(dropna=False))

Transaction channels:
channel
mobile     6414
web        3761
ATM        1012
mobille      60
MOBILE       47
unknown      37
WEB          36
weeb         24
ATm           9
Name: count, dtype: int64

KYC tiers:
kyc_tier
standard    7996
enhanced    1846
low         1047
NaN          302
standrd       72
STANDARD      70
unknown       32
ENHANCED      17
enhancd       12
LOW            5
NAN            1
Name: count, dtype: int64


### Observation

The `channel` and `kyc_tier` columns contain spelling mistakes, inconsistent capital letters and placeholder values.

Examples from the `channel` column include:

- `mobile`, `MOBILE` and `mobille`
- `web`, `WEB` and `weeb`
- `ATM` and `ATm`
- `unknown`

Examples from the `kyc_tier` column include:

- `standard`, `STANDARD` and `standrd`
- `enhanced`, `ENHANCED` and `enhancd`
- `low` and `LOW`
- `unknown`
- `NAN`
- 302 actual missing values shown as `NaN`

These variations may refer to the same categories, but Python currently treats them as different values.

They will be corrected and standardised during Day 3. No category values will be changed during the Day 2 assessment.

## 6. Check Numerical Data Quality

### What question are my answering?

I want to identify numerical values that may be incorrect, impossible or unusually high.

Values such as negative transaction counts or risk scores above their expected range may be data-entry errors and could affect the machine-learning model.

In [12]:
numerical_range_check = df[
    [
        "fee",
        "ip_risk_score",
        "device_trust_score",
        "txn_velocity_1h",
        "txn_velocity_24h"
    ]
].agg(["min", "max"])

numerical_range_check

,fee,ip_risk_score,device_trust_score,txn_velocity_1h,txn_velocity_24h
min,-1.00,0.004,-0.100,-1,0
max,9999.99,1.200,0.999,8,11


### Observation

The numerical range check identified several values that may be incorrect or unrealistic.

- `fee` has a minimum value of -1.00. A transaction fee would not normally be negative.
- `fee` has a maximum value of 9,999.99, which appears unusually high and may be an outlier or data-entry error.
- `ip_risk_score` has a maximum value of 1.20. If this score is expected to range from 0 to 1, this value is outside the expected range.
- `device_trust_score` has a minimum value of -0.10. If this score is expected to range from 0 to 1, the negative value is invalid.
- `txn_velocity_1h` has a minimum value of -1. A transaction count should not normally be negative.
- `txn_velocity_24h` ranges from 0 to 11, which appears reasonable based on the current dataset.

These suspicious values will not be changed during Day 2. They will be investigated and handled during the Day 3 data-cleaning process.

## 7. Explore Transaction Data and Metadata Fields

### What question are my answering?

I want to understand the types of information available in the NovaPay dataset.

This is important because fraud detection depends on transaction, customer, device, location and risk information.

In [13]:
print("Total number of columns:", len(df.columns))

for column in df.columns:
    print(column)

Total number of columns: 26
transaction_id
customer_id
timestamp
home_country
source_currency
dest_currency
channel
amount_src
amount_usd
fee
exchange_rate_src_to_dest
device_id
new_device
ip_address
ip_country
location_mismatch
ip_risk_score
kyc_tier
account_age_days
device_trust_score
chargeback_history_count
risk_score_internal
txn_velocity_1h
txn_velocity_24h
corridor_risk
is_fraud


### Observation

The NovaPay dataset contains 26 columns covering different types of information.

#### Transaction Information

- `transaction_id`
- `timestamp`
- `source_currency`
- `dest_currency`
- `channel`
- `amount_src`
- `amount_usd`
- `fee`
- `exchange_rate_src_to_dest`

#### Customer Information

- `customer_id`
- `home_country`
- `kyc_tier`
- `account_age_days`
- `chargeback_history_count`

#### Device and Location Information

- `device_id`
- `new_device`
- `ip_address`
- `ip_country`
- `location_mismatch`
- `device_trust_score`

#### Risk and Behaviour Information

- `ip_risk_score`
- `risk_score_internal`
- `txn_velocity_1h`
- `txn_velocity_24h`
- `corridor_risk`

#### Fraud Label

- `is_fraud`

These fields provide information about the transaction, customer, device, location, previous behaviour and risk level.

The `is_fraud` column is the target variable used to identify whether a transaction is legitimate or fraudulent.

## 8. Investigate a Potential Bias Source: Geography

### What question are my answering?

I want to examine whether fraud labels are distributed differently across customers from different home countries.

Geography may help identify unusual transaction patterns, but it could also create unfair predictions if the model learns to treat customers from a particular country as suspicious simply because of their location.

In [14]:
country_fraud_summary = (
    df.groupby("home_country")["is_fraud"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

country_fraud_summary.columns = [
    "Home Country",
    "Total Transactions",
    "Fraud Transactions",
    "Fraud Rate"
]

country_fraud_summary["Fraud Rate (%)"] = (
    country_fraud_summary["Fraud Rate"] * 100
).round(2)

country_fraud_summary[
    [
        "Home Country",
        "Total Transactions",
        "Fraud Transactions",
        "Fraud Rate (%)"
    ]
].sort_values(
    "Fraud Rate (%)",
    ascending=False
)

,Home Country,Total Transactions,Fraud Transactions,Fraud Rate (%)
0,CA,1228,185,15.07
1,UK,2132,282,13.23
2,US,8008,529,6.61
3,unknown,32,1,3.12


### Observation

The fraud rate differs across the customer home countries in the dataset.

- Canada has 1,228 transactions, with 185 labelled as fraud, giving a fraud rate of 15.07%.
- The United Kingdom has 2,132 transactions, with 282 labelled as fraud, giving a fraud rate of 13.23%.
- The United States has 8,008 transactions, with 529 labelled as fraud, giving a fraud rate of 6.61%.
- The `unknown` category has 32 transactions, with 1 labelled as fraud, giving a fraud rate of 3.12%.

These differences show that geography could influence the model.

However, a higher fraud rate for a country does not prove that customers from that country are naturally more fraudulent. The difference may be caused by sample size, transaction behaviour, previous investigation practices or bias in the historical fraud labels.

NovaPay should therefore use geographical information carefully and monitor whether the model unfairly flags customers from particular countries.

## 9. Investigate a Potential Bias Source: KYC Completeness

### What question are we answering?

We want to examine whether fraud rates differ across KYC tiers and whether transactions with missing KYC information are labelled differently.

KYC information may help NovaPay understand customer risk, but incomplete KYC records could cause the model to unfairly treat some customers as more suspicious.

In [15]:
kyc_fraud_summary = (
    df.assign(
        kyc_group=df["kyc_tier"].fillna("Missing")
    )
    .groupby("kyc_group")["is_fraud"]
    .agg(["count", "sum", "mean"])
    .reset_index()
)

kyc_fraud_summary.columns = [
    "KYC Tier",
    "Total Transactions",
    "Fraud Transactions",
    "Fraud Rate"
]

kyc_fraud_summary["Fraud Rate (%)"] = (
    kyc_fraud_summary["Fraud Rate"] * 100
).round(2)

kyc_fraud_summary[
    [
        "KYC Tier",
        "Total Transactions",
        "Fraud Transactions",
        "Fraud Rate (%)"
    ]
].sort_values(
    "Fraud Rate (%)",
    ascending=False
)

,KYC Tier,Total Transactions,Fraud Transactions,Fraud Rate (%)
7,low,1047,536,51.19
8,standard,7996,409,5.12
10,unknown,32,1,3.12
2,Missing,302,8,2.65
6,enhanced,1846,41,2.22
4,STANDARD,70,1,1.43
9,standrd,72,1,1.39
1,LOW,5,0,0.00
0,ENHANCED,17,0,0.00
5,enhancd,12,0,0.00


### Observation

The KYC results show that fraud rates differ across KYC groups.

- The `low` KYC tier has 1,047 transactions, with 536 labelled as fraud, giving a fraud rate of 51.19%.
- The `standard` tier has 7,996 transactions, with 409 labelled as fraud, giving a fraud rate of 5.12%.
- The `enhanced` tier has 1,846 transactions, with 41 labelled as fraud, giving a fraud rate of 2.22%.
- Transactions with missing KYC information have a fraud rate of 2.65%.
- The `unknown` category has a fraud rate of 3.12%.

The `low` KYC tier has a much higher fraud rate than the other main tiers, so it may be an important fraud indicator.

However, the KYC column also contains inconsistent labels such as `low` and `LOW`, `standard`, `STANDARD` and `standrd`, and `enhanced`, `ENHANCED` and `enhancd`.

These inconsistencies split the same category into separate groups and could distort the analysis. The labels must be standardised during Day 3 before drawing final conclusions.

NovaPay should also use KYC information carefully because incomplete or lower KYC status could cause some customers to be flagged more often. Model performance should be monitored to ensure that KYC-related decisions do not create unfair treatment.

## 10. Bias and Fairness in Financial Machine Learning

Bias can occur when a machine-learning model learns unfair patterns from historical data.

In fraud detection, historical fraud labels may reflect previous human decisions. If earlier investigations were inconsistent or unfair, the model may learn the same behaviour.

Attributes such as home country, IP country, KYC tier and missing KYC information could influence fraud predictions. These fields may be useful for identifying risk, but they could also cause some customer groups to be flagged more often.

False-positive fraud alerts can delay genuine transactions, block customers and create unnecessary manual investigations.

To reduce bias, NovaPay should:

- compare fraud predictions across different customer groups;
- monitor false-positive and false-negative rates;
- use explainable model predictions;
- allow fraud analysts to review important decisions;
- regularly check whether the model disadvantages particular groups.

## Reflection

### Which data attributes could skew fraud predictions?

Geography and KYC information could skew fraud predictions.

The home-country analysis showed different fraud rates across Canada, the United Kingdom and the United States. However, these differences may be caused by transaction behaviour, sample size or historical investigation practices rather than the country itself.

The KYC analysis showed that the `low` KYC tier had a much higher fraud rate than other groups. This could cause the model to treat all low-KYC customers as suspicious, even when their transactions are genuine.

Missing information, inconsistent category labels, chargeback history and historical fraud labels could also distort predictions.

NovaPay should therefore use these attributes carefully and monitor whether the model produces unfair results for particular customer groups.

# Day 2 Assessment Summary

## Main Findings

| Assessment Area | Finding |
|---|---|
| Dataset size | 11,400 transactions and 26 columns |
| Target column | `is_fraud` |
| Legitimate transactions | 10,403 — 91.25% |
| Fraudulent transactions | 997 — 8.75% |
| Class imbalance | Legitimate transactions are much more common than fraud transactions |
| Duplicate rows | 200 exact duplicates |
| Missing-value columns | 7 columns contain missing data |
| Category problems | Spelling errors, inconsistent capital letters and placeholder values |
| Numerical problems | Negative and unusually high values were identified |
| Chargeback history | Strong relationship with fraud, but possible data-leakage risk |
| Geography risk | Fraud rates differ across home countries |
| KYC risk | The low KYC tier has a much higher fraud rate |
| Bias concern | Geography, KYC status and historical fraud labels could skew predictions |

## Business Decision

The NovaPay dataset should not be used for machine-learning modelling in its current condition.

Before modelling begins, the following issues must be addressed:

- duplicate transactions;
- missing values;
- inconsistent category labels;
- suspicious numerical values;
- possible data leakage from chargeback history;
- potential bias connected with geography and KYC information.

No data was cleaned or deleted during Day 2. These issues will be handled during Week 1, Day 3: Data Cleaning and Preparation.

## Day 2 Conclusion

The Day 2 assessment helped us understand the data scope, fraud-label distribution, class imbalance, data-quality problems and potential bias risks.

The dataset contains useful fraud indicators, but it requires cleaning and further investigation before it can be used to train a reliable and fair machine-learning model.

# WEEK 1 – DAY 3

## Data Cleaning & Preparation

### Objective

To produce a clean, consistent and analysis-ready NovaPay transaction dataset while preserving the original fraud labels.

### Tasks / Activities

- Parse and standardise timestamps.
- Normalise text categories and correct inconsistent spellings.
- Remove exact duplicate records.
- Remove invalid or future-dated transactions.
- Handle the missing and invalid values identified during Day 2.
- Confirm that the fraud labels remain unchanged.
- Export the cleaned dataset as `cleaned_transactions.csv`.

### Expected Output

- Cleaned dataset: `cleaned_transactions.csv`
- Cleaning notebook: `03_cleaning.ipynb`
- Cleaning summary: `CLEANING_README.md`
- A before-and-after illustration of at least two cleaning steps

## 1. Create a Separate Copy for Cleaning

### What problem am I solving?

I need to preserve the original NovaPay dataset so that I can compare the raw data with the cleaned data.

I will load the original Excel file into `raw_df` and create a separate copy called `clean_df`.

All cleaning steps will be applied only to `clean_df`.

In [36]:
import pandas as pd
import numpy as np

raw_df = pd.read_excel("nova_pay_combined (1).xlsx")

clean_df = raw_df.copy()

print("Raw dataset shape:", raw_df.shape)
print("Cleaning copy shape:", clean_df.shape)

Raw dataset shape: (11400, 26)
Cleaning copy shape: (11400, 26)


## 2. Remove Exact Duplicate Transactions

### What problem am I solving?

The Day 2 assessment identified 200 exact duplicate rows.

Duplicate transactions can distort the analysis because the same transaction may be counted more than once. They may also give repeated records too much influence during model training.

I will remove only exact duplicate rows and keep the first occurrence of each transaction.

In [37]:
rows_before_duplicates = len(clean_df)
duplicates_before = clean_df.duplicated().sum()

clean_df = clean_df.drop_duplicates().copy()

rows_after_duplicates = len(clean_df)
duplicates_after = clean_df.duplicated().sum()

print("Rows before cleaning:", rows_before_duplicates)
print("Exact duplicates before cleaning:", duplicates_before)
print("Rows after cleaning:", rows_after_duplicates)
print("Exact duplicates after cleaning:", duplicates_after)
print("Rows removed:", rows_before_duplicates - rows_after_duplicates)

Rows before cleaning: 11400
Exact duplicates before cleaning: 200
Rows after cleaning: 11200
Exact duplicates after cleaning: 0
Rows removed: 200


### Before-and-After Observation

Before cleaning, the dataset contained 11,400 rows and 200 exact duplicate records.

After removing exact duplicates, the dataset contained 11,200 rows and no exact duplicate records.

This cleaning step improved data quality by ensuring that each identical transaction record is counted only once.

## 3. Parse and Standardise Timestamps

### What problem am I solving?

The Day 2 assessment found missing and invalid timestamp values.

A valid timestamp is important because transaction dates may later be used to examine fraud patterns over time.

I will convert the timestamp column into a proper date-and-time format and remove records with missing, invalid or future-dated timestamps.

### Check Missing and Invalid Timestamps

I will keep a copy of the original timestamp values before conversion.

This allows me to count the values that were already missing separately from the values containing invalid dates.

In [38]:
original_timestamps = clean_df["timestamp"].copy()

missing_timestamps = original_timestamps.isna().sum()

parsed_timestamps = pd.to_datetime(
    original_timestamps,
    errors="coerce",
    utc=True
)

invalid_timestamps = (
    original_timestamps.notna()
    & parsed_timestamps.isna()
).sum()

print("Missing timestamps:", missing_timestamps)
print("Invalid timestamps:", invalid_timestamps)

Missing timestamps: 29
Invalid timestamps: 31


### Convert the Timestamp Column

I will replace the original timestamp column with the converted UTC date-and-time values.

Invalid dates will appear as `NaT` and will be removed in the next cleaning step.

In [42]:
clean_df["timestamp"] = parsed_timestamps

print("Timestamp data type:", clean_df["timestamp"].dtype)

Timestamp data type: datetime64[ns, UTC]


### Check for Future-Dated Transactions

I will check whether any transaction date occurs after the current date because future-dated transactions may be incorrect.

In [43]:
current_date = pd.Timestamp.now(tz="UTC")

future_timestamps = (
    clean_df["timestamp"] > current_date
).sum()

print("Future-dated transactions:", future_timestamps)

Future-dated transactions: 0


### Remove Unusable Timestamps

I will remove transactions with missing, invalid or future-dated timestamps because they cannot be used reliably for time-based analysis.

In [41]:
rows_before = len(clean_df)

clean_df = clean_df[
    clean_df["timestamp"].notna()
    & (clean_df["timestamp"] <= current_date)
].copy()

rows_after = len(clean_df)

print("Rows before timestamp cleaning:", rows_before)
print("Rows after timestamp cleaning:", rows_after)
print("Rows removed:", rows_before - rows_after)

Rows before timestamp cleaning: 11200
Rows after timestamp cleaning: 11140
Rows removed: 60


### Before-and-After Observation

After duplicate removal, the dataset contained 11,200 rows.

The timestamp assessment identified 29 missing timestamps and 31 invalid timestamps. No future-dated transactions were found.

After removing the 60 records with unusable timestamps, 11,140 transactions remained.

The timestamp column was successfully converted from text into a consistent UTC date-and-time format. This improves data consistency and prepares the dataset for time-based analysis.


## 4. Normalise Text Categories and Correct Spellings

### What problem is being solved?

The Day 2 assessment identified inconsistent spelling, capitalisation and placeholder values in the `channel` and `kyc_tier` columns.

For example, `mobile`, `MOBILE` and `mobille` represent the same transaction channel, but Python currently treats them as different categories.

These values will be standardised to improve consistency before modelling.

### View the Categories Before Cleaning

The current channel and KYC values will be displayed before any changes are made. This provides a clear before-and-after comparison.

In [44]:
print("Channel values before cleaning:")
print(clean_df["channel"].value_counts(dropna=False))

print("\nKYC values before cleaning:")
print(clean_df["kyc_tier"].value_counts(dropna=False))

Channel values before cleaning:
channel
mobile     6253
web        3691
ATM         987
mobille      59
MOBILE       46
unknown      37
WEB          34
weeb         24
ATm           9
Name: count, dtype: int64

KYC values before cleaning:
kyc_tier
standard    7798
enhanced    1802
low         1038
NaN          297
standrd       72
STANDARD      69
unknown       30
ENHANCED      16
enhancd       12
LOW            5
NAN            1
Name: count, dtype: int64


### Standardise Channel and KYC Categories

The `channel` and `kyc_tier` columns will be converted to lowercase, extra spaces will be removed, spelling mistakes will be corrected and placeholder values will be treated as missing information.

In [53]:
clean_df["channel"] = (
    clean_df["channel"]
    .astype("string")
    .str.strip()
    .str.lower()
)

clean_df["channel"] = clean_df["channel"].replace({
    "mobille": "mobile",
    "weeb": "web",
    "unknown": pd.NA
})

In [50]:
clean_df["kyc_tier"] = (
    clean_df["kyc_tier"]
    .astype("string")
    .str.strip()
    .str.lower()
)

clean_df["kyc_tier"] = clean_df["kyc_tier"].replace({
    "standrd": "standard",
    "enhancd": "enhanced",
    "unknown": pd.NA,
    "nan": pd.NA
})

### View the Categories After Cleaning

The cleaned channel and KYC values will be displayed to confirm that inconsistent spellings and capitalisation have been corrected.

In [52]:
print("Channel values after cleaning:")
print(clean_df["channel"].value_counts(dropna=False))

print("\nKYC values after cleaning:")
print(clean_df["kyc_tier"].value_counts(dropna=False))

Channel values after cleaning:
channel
mobile    6358
web       3749
atm        996
<NA>        37
Name: count, dtype: Int64

KYC values after cleaning:
kyc_tier
standard    7939
enhanced    1830
low         1043
<NA>         328
Name: count, dtype: Int64


### Before-and-After Observation

Before cleaning, the `channel` and `kyc_tier` columns contained inconsistent capitalisation, spelling mistakes and placeholder values.

After cleaning, the transaction channels were reduced to the consistent categories `mobile`, `web` and `atm`.

The KYC values were reduced to the consistent categories `standard`, `enhanced` and `low`.

Placeholder values such as `unknown` and `NAN` were treated as missing information.

This cleaning step improved consistency and prevented the same category from being interpreted as several different values.


## 5. Handle Remaining Missing Values

### What problem is being solved?

Some columns still contain missing information after the earlier cleaning steps.

Numerical missing values will be replaced with the median because the median is less affected by unusually high or low values.

Text missing values will be labelled as `missing` so that the transactions can remain in the dataset.

In [54]:
remaining_missing = clean_df.isna().sum()

remaining_missing[
    remaining_missing > 0
]

channel                37
amount_usd            300
fee                   290
ip_address            300
ip_country            297
kyc_tier              328
device_trust_score    290
dtype: int64

### Replace Missing Numerical Values

Missing values in `amount_usd`, `fee` and `device_trust_score` will be replaced with the median value of each column.

In [56]:
numeric_columns = [
    "amount_usd",
    "fee",
    "device_trust_score"
]

for column in numeric_columns:
    clean_df[column] = clean_df[column].fillna(
        clean_df[column].median()
    )

In [57]:
print(clean_df[numeric_columns].isna().sum())

amount_usd            0
fee                   0
device_trust_score    0
dtype: int64


### Replace Missing Text Values

Placeholder values in the IP-country column will first be treated as missing.

Missing values in the text columns will then be labelled as `missing`.

In [58]:
clean_df["ip_country"] = (
    clean_df["ip_country"]
    .astype("string")
    .str.strip()
    .str.upper()
)

clean_df["ip_country"] = clean_df["ip_country"].replace({
    "UNKNOWN": pd.NA,
    "NAN": pd.NA
})

text_columns = [
    "channel",
    "ip_address",
    "ip_country",
    "kyc_tier"
]

for column in text_columns:
    clean_df[column] = clean_df[column].fillna("missing")

### Confirm the Missing Values Were Handled

The dataset will now be checked again to confirm whether any missing values remain.

In [59]:
print(
    "Total missing values after cleaning:",
    clean_df.isna().sum().sum()
)

Total missing values after cleaning: 0


### Observation

The remaining numerical missing values were replaced with the median of their respective columns.

Missing text information was labelled as `missing` rather than deleting the affected transactions.

After this step, the dataset contained no remaining missing values.

### Replace Missing Text Values

Missing values in the text columns will be labelled as `missing` so that the affected transactions can remain in the dataset.

In [60]:
text_columns = [
    "channel",
    "ip_address",
    "ip_country",
    "kyc_tier"
]

for column in text_columns:
    clean_df[column] = clean_df[column].fillna("missing")

In [61]:
print(clean_df[text_columns].isna().sum())

channel       0
ip_address    0
ip_country    0
kyc_tier      0
dtype: int64


### Confirm All Missing Values Were Handled

The cleaned dataset will now be checked to confirm whether any missing values remain.

In [62]:
print(
    "Total missing values after cleaning:",
    clean_df.isna().sum().sum()
)

Total missing values after cleaning: 0


### Observation

Missing numerical values were replaced with the median of their respective columns.

Missing text values were labelled as `missing` rather than deleting the affected transactions.

After this step, no missing values remained in the cleaned dataset.

In [63]:
 print(
    "Total missing values after cleaning:",
    clean_df.isna().sum().sum()
)

Total missing values after cleaning: 0


### Observation

The remaining numerical missing values were replaced with the median of their respective columns.

Missing text information was labelled as `missing` rather than deleting the affected transactions.

After this step, the dataset contained no remaining missing values.

## 6. Correct Invalid Numerical Values

### What problem is being solved?

The Day 2 assessment identified numerical values that are impossible or outside their expected ranges.

These include negative transaction amounts, invalid fees, risk scores outside 0 to 1, negative device-trust scores and negative transaction counts.

In [64]:
invalid_numeric_summary = pd.Series({
    "Invalid fee values": (
        (clean_df["fee"] < 0)
        | (clean_df["fee"] == 9999.99)
    ).sum(),

    "Invalid source amounts": (
        clean_df["amount_src"] <= 0
    ).sum(),

    "Invalid IP risk scores": (
        (clean_df["ip_risk_score"] < 0)
        | (clean_df["ip_risk_score"] > 1)
    ).sum(),

    "Invalid device trust scores": (
        (clean_df["device_trust_score"] < 0)
        | (clean_df["device_trust_score"] > 1)
    ).sum(),

    "Invalid one-hour transaction counts": (
        clean_df["txn_velocity_1h"] < 0
    ).sum()
})

invalid_numeric_summary


Invalid fee values                     200
Invalid source amounts                 100
Invalid IP risk scores                 200
Invalid device trust scores            200
Invalid one-hour transaction counts    200
dtype: int64

### Replace Invalid Numerical Values

Invalid continuous values will be replaced with the median of the valid values in their column.

Negative one-hour transaction counts will be replaced with zero because a transaction count cannot be negative.

In [66]:
valid_fee_median = clean_df.loc[
    (clean_df["fee"] >= 0)
    & (clean_df["fee"] != 9999.99),
    "fee"
].median()

clean_df.loc[
    (clean_df["fee"] < 0)
    | (clean_df["fee"] == 9999.99),
    "fee"
] = valid_fee_median

In [67]:
valid_amount_median = clean_df.loc[
    clean_df["amount_src"] > 0,
    "amount_src"
].median()

clean_df.loc[
    clean_df["amount_src"] <= 0,
    "amount_src"
] = valid_amount_median

In [68]:
valid_ip_risk_median = clean_df.loc[
    clean_df["ip_risk_score"].between(0, 1),
    "ip_risk_score"
].median()

clean_df.loc[
    ~clean_df["ip_risk_score"].between(0, 1),
    "ip_risk_score"
] = valid_ip_risk_median

In [69]:
valid_device_score_median = clean_df.loc[
    clean_df["device_trust_score"].between(0, 1),
    "device_trust_score"
].median()

clean_df.loc[
    ~clean_df["device_trust_score"].between(0, 1),
    "device_trust_score"
] = valid_device_score_median

In [70]:
clean_df.loc[
    clean_df["txn_velocity_1h"] < 0,
    "txn_velocity_1h"
] = 0

### Confirm the Invalid Values Were Corrected

In [71]:
print(
    "Invalid fees:",
    (
        (clean_df["fee"] < 0)
        | (clean_df["fee"] == 9999.99)
    ).sum()
)

print(
    "Invalid source amounts:",
    (clean_df["amount_src"] <= 0).sum()
)

print(
    "Invalid IP risk scores:",
    (~clean_df["ip_risk_score"].between(0, 1)).sum()
)

print(
    "Invalid device trust scores:",
    (~clean_df["device_trust_score"].between(0, 1)).sum()
)

print(
    "Invalid one-hour transaction counts:",
    (clean_df["txn_velocity_1h"] < 0).sum()
)

Invalid fees: 0
Invalid source amounts: 0
Invalid IP risk scores: 0
Invalid device trust scores: 0
Invalid one-hour transaction counts: 0


### Observation

The invalid numerical values were corrected using clear validation rules.

Invalid continuous values were replaced with the median of valid records, while negative transaction counts were replaced with zero.

After cleaning, no invalid values remained in the checked numerical columns.

## 7. Final Data Quality Check

The cleaned dataset will be checked to confirm that duplicates, missing values and invalid fraud labels are no longer present.

In [72]:
print("Final dataset shape:", clean_df.shape)
print("Remaining duplicates:", clean_df.duplicated().sum())
print("Remaining missing values:", clean_df.isna().sum().sum())
print("Missing fraud labels:", clean_df["is_fraud"].isna().sum())
print("Fraud label values:", sorted(clean_df["is_fraud"].unique()))

Final dataset shape: (11140, 26)
Remaining duplicates: 0
Remaining missing values: 0
Missing fraud labels: 0
Fraud label values: [np.int64(0), np.int64(1)]


### Fraud Label Integrity

The fraud labels were checked after cleaning.

The target column still contains only `0` and `1`, and no fraud labels are missing.

This confirms that the cleaning process did not change or damage the target variable.

## 8. Export the Cleaned Dataset

The final cleaned dataset will be saved as `cleaned_transactions.csv` for the next stage of the project.

In [73]:
clean_df.to_csv(
    "cleaned_transactions.csv",
    index=False
)

print("cleaned_transactions.csv created successfully.")


cleaned_transactions.csv created successfully.
